### PR TITLE
Provide a fast path for ReadCallerFrame when the caller is a #send.

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java [hunk 1]
```diff
@@ -60,10 +63,17 @@ public CachedDispatchNode(
...
         CompilerAsserts.neverPartOfCompilation("Dispatch nodes should not be altered after compilation.");
-        if (!needsCallerFrame) {
-            needsCallerFrame = true;
+        if (!sendsFrame) {
+            sendsFrame = true;
+        }
+    }
+
+    public void replaceSendingCallerFrame(CallerFrameAccess access) {
+        CompilerAsserts.neverPartOfCompilation("Dispatch nodes should not be altered after compilation.");
```
Comment (chrisseaton) [inline truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java]: I missed this before. What exactly does it mean? What happens if methods are monkey patched?

Comment (aardvark179) [inline truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java]: You're right. The ReadCallerFrameNode should not try and change things if it has been compiled, The dispatch nodes should revert to the interpreter and invalidate.

File: truffleruby/src/main/java/org/truffleruby/language/arguments/ReadCallerFrameNode.java [hunk 2]
```diff
@@ -63,7 +64,7 @@ private Frame getCallerFrame() {
         if (!CompilerDirectives.inCompiledCode()) {
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/arguments/ReadCallerFrameNode.java]: inCompiledCode() will always be `false` behind a boundary.

File: truffleruby/src/main/java/org/truffleruby/language/arguments/ReadCallerFrameNode.java [hunk 3]
```diff
@@ -46,13 +45,15 @@ public Frame execute(VirtualFrame frame) {
...
 
     private void replaceDispatchNode() {
-        CompilerAsserts.neverPartOfCompilation("Dispatch nodes should never be replaced after compilation.");
-        if (!getContext().getCallStack().callerIsSend()) {
-            Node callerNode = getContext().getCallStack().getCallerNode();
-            if (callerNode instanceof DirectCallNode) {
-                Node parent = callerNode.getParent();
-                if (parent instanceof CachedDispatchNode) {
-                    ((CachedDispatchNode) parent).replaceSendingChild();
+        CompilerAsserts.neverPartOfCompilation("No fast path frame access changes after compilation.");
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/arguments/ReadCallerFrameNode.java]: No need for this assert, it's called behind a boundary.

**Pull request merged**